### ISSUE TITLE
add hasTurnCosts to Weighting

### ISSUE THREAD

**Issue by karussell** [graphhopper/graphhopper#1965]
We need to fix #1960 in a different way to avoid issues with e.g. customized GraphHopper classes that override createWeighting.

**Comment by easbar** [comment]
I'm not sure about fixing #1960 in a different way, I think I would rather add 
```java
boolean hasTurnCosts() {
    return turnCostProvider == NO_TURN_COST_PROVIDER;
}
```
to Weighting and then use this check in Landmarkstorage to make sure the weighting is not using turn costs (right now this error will go unnoticed)

**Comment by karussell** [comment]
Ah, sure! Can we remove the tests as we would get an exception then? (I know in general tests cannot hurt but they mean execution time and migration time)

**Comment by easbar** [comment]
The tests should still be fine even when adding the check. I think the test that compares random requests between the algorithms is useful either with or without issue #1960 (and worth the execution time). The one with the fixed values, well I would also keep it (to make sure we are not in a situation where *all* algos went wrong)

**Comment by easbar** [comment]
And I think in general tests **can** hurt :)

**Comment by karussell** [comment]
Ok

> The one with the fixed values, well I would also keep it (to make sure we are not in a situation where all algos went wrong)

Isn't this covered from the many other tests?

**Comment by easbar** [comment]
No, because they usually focus on the algorithms and do not use the `GraphHopper` class. The test we are talking about here is even using route resource.

**Comment by easbar** [comment]
Ah sorry you meant the one with the fixed values? Yes you are right that one isn't so special and we can probably remove it.

**Comment by easbar** [comment]
Thinking about this again I really think the issue we are having here is that the algorithms do not consider the traversal mode when calculating the weights. That is why we now have to be careful not to pass a weighting that calculates turn costs into a graph algorithm that runs node-based. We noticed this here for LM #1960, but its also the same for isochrones and spt: #1967. 

There is no `prevOrNextEdgeId` parameter in `Weighting` anymore. The only thing we can do with a `Weighting` is to calculate edge and turn weights separately. We have a utility method `GHUtility#calcWeightWithTurnWeight` that calculates the sum of edge and turn weights and thus has a `prevOrNextEdgeId` parameter, but I see no use case that calls this method in a meaningful way when running node-based. Even before #1863 where we made this change we were passing NO_EDGE sometimes to indicate we do not want turn costs. Similarly I think all we have to do here is exclude the turn weight calculation if `!traversalMode.isEdgeBased()`. The alternative would be to  make sure that either the traversal mode is edge-based or the weighting does not support turn costs whenever we create a routing algorithm object. But what could we do this waỳ that we cannot do when we explicitly exclude the turn weights in the algorithms?

Hm and what is more ugly: We cannot use a weighting with turn costs with algorithms running node based or the turn costs are ignored when running an algorithm node-based? (I'm not sure tbh)
